### PR TITLE
HDDS-9173. Invalidate snapshot cache once snapshot gets purged

### PR DESCRIPTION
**Pull request by DaveTeng0**
## What changes were proposed in this pull request?
Invalidate snapshot cache once snapshot gets purged

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9173

## How was this patch tested?
Integration test

### CODE REVIEW

**Comment by aswinshakil** [comment]
@DaveTeng0 Can you fix the checkstyle?

**Comment by DaveTeng0** [comment]
> @DaveTeng0 Can you fix the checkstyle?

oh sure! Thanks!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotPurgeRequest.java [hunk 1]
```diff
@@ -103,6 +103,8 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
             trxnLogIndex, updatedSnapInfos, true);
         updateSnapshotChainAndCache(omMetadataManager, fromSnapshot,
             trxnLogIndex, updatedPathPreviousAndGlobalSnapshots);
+        ozoneManager.getOmSnapshotManager().getSnapshotCache()
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotPurgeRequest.java:106]: Can you please add test for it? Could be as simple as checking cache size after the purge finishes.

Comment (DaveTeng0) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotPurgeRequest.java:106]: Yup! sure definitely! @hemantk-12 please take a look the new verification part added in [TestSnapshotDeletingService.java](https://github.com/apache/ozone/pull/5248/commits/f7a4b3886cd67fa7ba1275aca9a70197bdab88ff#diff-4467e2d5f908fe0f61428c3328bd0a0c797d0b07e3a38ec40969f2264277ecba)
Feel free to let me know if any question, thanks!

Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotPurgeRequest.java:106]: Thanks @DaveTeng0

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotDeletingService.java [hunk 2]
```diff
@@ -188,6 +191,30 @@ public void testMultipleSnapshotKeyReclaim() throws Exception {
     assertTableRowCount(deletedTable, 0);
 
     verifySnapshotChain(delSnapInfo, null);
+
+    // verify the cache of purged snapshot
+    // /vol1/bucket2/bucket2snap1 has been cleaned up from cache map
+    SnapshotCache snapshotCache = om.getOmSnapshotManager().getSnapshotCache();
+    Method getDbMapMethod = snapshotCache.getClass()
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotDeletingService.java:198]: I think you can directly use [SnapshotCache#size()](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java#L94).

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotDeletingService.java [hunk 3]
```diff
@@ -188,6 +191,30 @@ public void testMultipleSnapshotKeyReclaim() throws Exception {
...
+    getDbMapMethod.setAccessible(true);
+    ConcurrentHashMap<String,
+        ReferenceCounted<IOmMetadataReader, SnapshotCache>> dbMap =
+        (ConcurrentHashMap<String,
+            ReferenceCounted<IOmMetadataReader,
+                SnapshotCache>>) getDbMapMethod.invoke(snapshotCache);
+
+    assertEquals(2, dbMap.size());
+
+    Method getPendingEvictionListMethod =
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotDeletingService.java:209]: Similar to previous comment, may be create a public function for `pendingEvictionList#size`.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotDeletingService.java (hunk 2 above)
Comment (DaveTeng0) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotDeletingService.java:198]: oh! you're absolutely right! let me change it!

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotDeletingService.java (hunk 3 above)
Comment (DaveTeng0) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotDeletingService.java]: yeah, that makes sense! updated! Thanks @hemantk-12 !

**Review by hemantk-12 — APPROVED** [review]
LGTM.

Thanks for the patch @DaveTeng0

**Pull request merged**